lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp

```diff
@@ -0,0 +1,472 @@
+          }
+        }
+      }
+    }
+  }
+  dummyNode->clearChilds();
+
+  // Collect all inputs of the module and a port direction list.
+  SmallVector<BlockArgument, 4> inputs;
+  SmallVector<bool, 8> directionList;
+  unsigned index = 0;
+  for (auto port : module.getPorts()) {
+    directionList.push_back(port.isOutput());
+    if (!port.isOutput())
+      inputs.push_back(module.getPortArgument(index));
+    ++index;
+  }
+
+  // Launch a DFS for each input to find outputs that are combinationally
+  // connected to it.
+  for (auto input : inputs) {
+    auto inputIndex = input.getArgNumber();
+    for (auto node : llvm::depth_first<CombNode>(combGraph.getNode(input))) {
+      if (auto nodeVal = node->getValOrOp().dyn_cast<Value>()) {
+        if (auto output = nodeVal.dyn_cast<BlockArgument>()) {
+          auto outputIndex = output.getArgNumber();
+          if (directionList[outputIndex])
+            combPathsMap[module][inputIndex].push_back(outputIndex);
+        }
+      }
+    }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Hashing CombNode
+//===----------------------------------------------------------------------===//
+
+namespace llvm {
+template <>
+struct DenseMapInfo<CombNode> {
+  static CombNode getEmptyKey() {
+    auto pointer = llvm::DenseMapInfo<void *>::getEmptyKey();
+    return static_cast<CombNode>(pointer);
+  }
+  static CombNode getTombstoneKey() {
+    auto pointer = llvm::DenseMapInfo<void *>::getTombstoneKey();
+    return static_cast<CombNode>(pointer);
+  }
+  static unsigned getHashValue(const CombNodeImpl *node) {
+    return node->hash_value();
+  }
+  static bool isEqual(const CombNodeImpl *lhs, const CombNodeImpl *rhs) {
+    auto empty = getEmptyKey();
+    auto tombstone = getTombstoneKey();
+    if (lhs == empty || lhs == tombstone || rhs == empty || rhs == tombstone)
+      return lhs == rhs;
+    return *lhs == *rhs;
+  }
+};
+} // namespace llvm
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// This pass constructs a local graph for each module to detect combinational
+/// cycles. To capture the cross-module combinational cycles, this pass inlines
+/// the combinational paths between IOs of its subinstances into a subgraph and
+/// encodes them in a `combPathsMap`.
+class CheckCombCyclesPass : public CheckCombCyclesBase<CheckCombCyclesPass> {
+  void runOnOperation() override {
+    auto &instanceGraph = getAnalysis<InstanceGraph>();
+    // Traverse modules in a post order to make sure the combinational paths
+    // between IOs of a module have been detected and recorded in `combPathsMap`
+    // before we handle its parent modules.
+    for (auto node : llvm::post_order<InstanceGraph *>(&instanceGraph)) {
+      // TODO: Handle FExtModuleOp with annotations.
```

Comment:
Why the todo?  ExtModule's cannot have cycles since they have no body.

Reply:
😅  So, there are these [`CombinationalPath` Annotations](https://www.chisel-lang.org/api/firrtl/latest/firrtl/transforms/CombinationalPath.html) which can be used to inform the compiler about a combinational path in a blackbox. This is what Hanchen is referring to, here. CIRCT can't pass all SFC tests if it doesn't support this, but I'm fine to lazily implement this once there's an obvious user.